text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,639 @@
+| Plaintext                  | 23k         | -             |
+| gzipped                    | 3.6k        | 4.8k          |
+| Zopfli                     | 3.4k        | 4.6k          |
+| bzip2                      | 3.4k        | 4.6k          |
+| Prefix-grouped (plaintext) | 7.4k        | -             |
+| Prefix-grouped gzipped     | 3.0k        | 4.0k          |
+| Bloom Filter (1% error)    | 720b        | 960b          |
+
+Prefix-grouped, gzipped data seems to give the best results that are still
+convenient and generic to work with (see Appendix 2 for a description of
+prefix grouping).
+
+See Appendix 3 for an example blob to be added to a template. I'm not super
+happy with this, but this is the most straightfoward option available.
+
+If we are willing to do more work, we can achieve less space used up by doing
+more esoteric things. One option would be to use a Bloom Filter, which can use
+a configurable amount of storage to achieve a configurable hit error ratio.
+For example, achieving about a 1% false positive rate would take on the order
+of about ~1k base64-encoded bytes. However, in order to extract a construct list
+from this Bloom Filter would be a considerable amount of work: we would need
+a full list of all construct available in every library and at every version,
+in order to test each construct's presence in the filter. This would destroy
+the openness of the protocol, and is probably not worth it.
+
+### Analytics payload
+
+The analytics payload itself will be preceded by an encoding scheme identifier,
+followed by a `:` and the transmitted data:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Analytics: <payload-encoding>:<payload>
+```
+
+The only valid payload encodings at the moment are:
+
+* `deflate64`: base64-encoded, gzipped data.
+* `plain`: plaintext data.
+
+The payload itself is a JSON object with `"constructs"` currently defined as the only
+valid key (but can be extended in the future). Values can be prefixed with
+an encoding scheme as well:
+
+```json
+{
+  "constructs": "[<value-encoding>:]<prefix-encoded construct list>"
+}
+```
+
+The only valid payload value encodings at the moment are:
+
+* `prefix`: prefix-encoded string list (see Appendix 2).
+* `plain`: plaintext data.
+
+Encoding may change the type (the encoded value is a `string` but the decoded
+value is a `string[]`).
+
+> If a value does not start with > /^[a-z0-9]{,10}:/ it is considered to
+> be not encoded (the encoding is considered to be "plain").
+
+## Metadata resource decoding
+
+Decoding will happen on the receiving end before the data hits long-term storage,
+in order to make consuming it as trivial as possible. There is no need to be frugal
+with bits anymore once the data hits an AWS datacenter.
+
+# Extensions
+
+We will be able more information into additional fields of the metadata resource.
+Size of template, number of resources are obvious candidates.
+
```

Comment:
There needs to be a section around end-to-end verification. 

How will we know when the data collection logic has an error (not counting or mis-counting), but none of our unit tests fail?

Reply:
Please address this comment. 

Are we confident all parts of the system can be tested? Could we have a situation where metrics are not counted and we're flying blind?

How about e2e tests/canaries from specific accounts?